### PR TITLE
Auth: Refactor BaseClient._get_creds to read from config even if creds are set

### LINKED ISSUE #6846: x509 auth: client does not read from configuration file whe creds are passed
### Description

From https://github.com/rucio/rucio/pull/6838#issuecomment-2190913409:

> This PR introduced an annoying regression to `x509` authentication. The client reads from the configuration file only when `creds` is unset.
> 
> https://github.com/rucio/rucio/blob/0bbc069faf17f06eff1891aa90e634adea67b216/lib/rucio/client/baseclient.py#L264-L266

### Steps to reproduce

Attempt x509 auth with creds passed

### Rucio Version

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by rdimaio**
Fix #6846

### CODE REVIEW

**Comment by maxnoe** [comment]
Can we refactor this a bit further so that

* All options have a clear precedence of (from lowest to highest priority):
    * default value
    * configuration file
    * environment variable
    * `creds` argument

* take care that when checking the configuration file, no `No such section` error is thrown?
   I think for this, it might be easiest if the `config_get(section, name, default)` function would just return `default` in case the section doesn't exist instead of raising an error.


When using the cli or inside a docker container, it would be nice if rucio wouldn't need a configuration file at all when everything is defined either as arguments or via env variables. 


As far as I can tell there are a couple of options for which there are neither env variables nor cli options, just the configuration file value.

**Comment by rdimaio** [comment]
> Can we refactor this a bit further so that
> 
>     * All options have a clear precedence of (from lowest to highest priority):
>       
>       * default value
>       * configuration file
>       * environment variable
>       * `creds` argument
> 
>     * take care that when checking the configuration file, no `No such section` error is thrown?
>       I think for this, it might be easiest if the `config_get(section, name, default)` function would just return `default` in case the section doesn't exist instead of raising an error.
> 
> 
> When using the cli or inside a docker container, it would be nice if rucio wouldn't need a configuration file at all when everything is defined either as arguments or via env variables.
> 
> As far as I can tell there are a couple of options for which there are neither env variables nor cli options, just the configuration file value.

I agree, but I think it might be best to do this in a follow-up PR so that we can merge your changes from https://github.com/rucio/rucio/pull/6848 first

**Comment by maxnoe** [comment]
@rdimaio I wasn't yet totally sure how to implement #6848 properly, i.e. the error handling in case something is empty or not valid. So it has a large overlap.

I thought maybe doing this on one go, as one has to basically touch all lines here in this _get_creds function might be easiest.

File: lib/rucio/client/baseclient.py [hunk 1]
```diff
@@ -239,57 +239,62 @@ def _get_auth_type(self, auth_type: Optional[str]) -> str:
...
+        if not creds:
             self.logger.debug('No creds passed. Trying to get it from the config file.')
             creds = {}
-            try:
-                if self.auth_type in ['userpass', 'saml']:
+
+        try:
+            if self.auth_type == 'oidc':
+                # if there are default values, check if rucio.cfg does not specify them, otherwise put default
+                if 'oidc_refresh_lifetime' not in creds or creds['oidc_refresh_lifetime'] is None:
```
Comment (voetberg) [inline lib/rucio/client/baseclient.py:249]: All of these if "key" not in creds or creds['key'] is None could be replaced with 

```
if creds.get('key') is None: 
```

For legibility

Comment (voetberg) [reply lib/rucio/client/baseclient.py:249]: This whole oidc auth block could also probably be a for loop, just looping through a list of the keys - A debug log would be able to track where errors happened in this case. The whole block is just a lot to read.

**Review by voetberg — CHANGES_REQUESTED** [review]
I largely agree with @maxnoe 's comments here - There should be a clear logic for order of preference on the auth variables (I was having a hard time with #6839 for this exact reason). 

Maybe a better solution would be to have a smaller helper function that went through and checked each place a single auth variable can be found in order to keep _get_creds form exploding into hydra of a function

**Comment by bari12** [comment]
Should we split this up into two different issues? I would like to fix the regression in the upcoming release.

**Comment by rdimaio** [comment]
> Should we split this up into two different issues? I would like to fix the regression in the upcoming release.

I think it would be best to merge @maxnoe's PR first: https://github.com/rucio/rucio/pull/6848 - I can then adapt my PR to that one and address @voetberg 's comments

**Comment by bari12** [comment]
I am merging this now, because the fix should be included in the 35.0.0 release. But please let's continue to include the other points mentioned in @maxnoe PR #6848 for the following release.

**Review by bari12 — APPROVED** [review]

**Pull request merged**